### PR TITLE
extract styles to individual css files

### PR DESCRIPTION
**Pull request by dsantos**
One step towards tidying up the html files and avoid duplication is to organise the styles. 

The idea is to extract all styles as-is from each html file and put them in individual css files which are linked by the corresponding pages. Then we can identify common styles and put them in one file linked by all pages, leaving each page's specific styles in individual files. What do you think about it?

Also, there's a lot of inline styling that ideally should be defined as classes in these files. Better than that would be to have a css framework like bootstrap, material, tailwind, etc. - don't know if you ever considered it.

### CODE REVIEW

File: static/views/companies.html [hunk 1]
```diff
@@ -453,7 +434,7 @@ <h4 style="margin-left: 9px;">Golang Cafe</h4>
 			alert('Please enter a valid email address');
 			return;
 		}
-		postMessage({email: emaile}, function(status) {
+		postMessage({email, message}, function(status) {
```
Comment (dsantos) [inline static/views/companies.html:437]: don't think this ever worked 🤔

**Comment by dm161** [comment]
@dsantos thank you for the PR - the reason why everything is inlined is because of performance. usability and speed for people using the site is a key metric i want to keep optimising for. So pulling out styles would be a step in the opposite direction.
I have done dozens of fullstack side projects and in my experience I found front-end frameworks to be overwhelming and with little value. Because they introduce a lot of unnecessary complexity and slow down the site and development speed (compiling css, bundling up, slow down all the local development workflow, more complex deployment). Also this will require a complete UI refactoring of the site as all the style will be out of whack

I understand it's frustrating to update the HTML views and I was planning to do the following to make everything much cleaner.

The first step towards better maintenance and developer experience for this would be to keep the styles inlined but having nested templates. This will allow us to split up these big html views into small subcomponents which are merged together using Go nested template feature. Conceptually speaking like react components (reusable components that have html/css/js in one place and can be reused but much more simple)

So we would get a footer.html template which includes only html/css/js used for the footer for instance. A header.html template with only header html/css/js. A navbar.html and so on and so forth... There is an issue tracking this objective #36 

What do you think of the nested templates? will this not solve lots of the bloating of these views?

File: static/views/companies.html (hunk 1 above)
Comment (dm161) [reply static/views/companies.html:437]: good catch.. some of these pages have lots of copy pasted scripts. So just follow my PR comment I left. would be beneficial to use nested templates and split all re-usable html/css/js in sub-templates so we won't have to copy-paste the same stuff over and over again :)

**Comment by dsantos** [comment]
@dm161 thank you for your comment

To clarify, when I said "inline styling" I meant `style="..."`. Nonetheless, I don't think it is totally true that helps with performance, mainly because:
- the time it takes one request to retrieve a page (html + styles) VS two parallel requests, one for the html and other for the styles might turn out to be the same
- if you choose to return html + styles in every request it's possible that you're not leveraging browser caching

Regarding css frameworks, I've used them in the past and never had big issues, however have to admit they were used since the beginning of the project. It's also a matter of preference. 

Regarding templates, I agree they can and will help composing the website by splitting it in different sections and reduce duplication. But, the concept of html/css/js in each section violates at least these conventions: 
- the `<style>` element should be inside the `<head>` element of the document (unless you intend to use `style="..."` everywhere)
- have the scripts at the bottom of the body to ensure they're fetched/applied after the DOM's loaded (unless not necessary)

Unless I misunderstood your idea?

**Comment by dm161** [comment]
- yes we should get rid of inline css ie. `style="font-size:.."` on each html node and move that to internal css blocks 👍
- using external css is a no-go - external css style are a blocking request and will increase FCP latency. Once the browser reads the style external directive it will need to stop rendering the page until the CSS has been retrieved. This will take a while more because of all the connection round trips to connect to the server again and download the CSS. You might have an argument if you say the css gets cached after the first request, but so does html+css as it's currently, the page with all the internal css is under 150kb. I would try to avoid putting the css as external resource unless there are very strong reasons to do so
- your point about having css only inside the head is correct. 👍 We can still split up css/js/html into different file templates for each category and include it using Go nested templates.
So like you did with your `companies.css`, `developers.css` but this will be html files only containing html code (with inside css or js) and will be included in the right location. In the case of css the template will be rendered in the header. for example

**header-css.html**
```
{{ define "header-css" }}
<style type="text/css">...</style>
{{ end }}
```
**footer-css.html**
```
{{ define "footer-css" }}
<style type="text/css">...</style>
{{ end }}
```
**developers.html**
```
<head>
...
{{ template "header-css" }}
{{ template "footer-css" }}
</head>
...
```

- same applies for javascript, you raised a good point that it should be at the end of the body 👍 and we can definitely split up the javascript/css/html for each template into different files and include these components individually in the right locations, so for developer js code it will be as follows

**header-js.html**
```
{{ define "header-js" }}
<script>...</script>
{{ end }}
```
**footer-js.html**
```
{{ define "footer-js" }}
<script>...</script>
{{ end }}
```

**developers.html**
```
<head>
...
{{ template "header-css" }}
{{ template "footer-css" }}
</head>
<body>
...
{{ template "header-js" }}
{{ template "footer-js" }}
</body>
```

**Comment by dsantos** [comment]
I see that you already made up your mind about not using external css and its fine because works either way. For curiosity I suggest you measure the performance of both approaches to see if the benefit outweighs the cost. Unfortunately, the current method its harder to review and maintain because of being minified and duplicated.

In your example above, you're still be repeating `<head>`, `<body>`... for each page. I think it's possible to have a single html layout for the website and each page to set and override parts of it. Something along these lines:

```
<html>
<head>
...
<style>common</style>
{{ template "extended-css" }}
</head>
<body>
{{ template "header" }}
{{ template "footer" }}
<script>common</script>
{{ template "extended-js" }}
</body>
</html>
```

or provide more flexibility in the composition for each page:

```
<html>
<head>
...
<style>common</style>
{{ template "extended-css" }}
</head>
<body>
{{ template "page" }}
<script>common</script>
{{ template "extended-js" }}
</body>
</html>
```

**Comment by dm161** [comment]
Great thank you for your comments @dsantos - Yes my example was the first iteration towards a fully re-usable html template. I think we should proceed by iterations as it will be easier to see the impact and easier to review

You are right with regards to maintainability having internal minified css copy-pasted everywhere. With the approach I proposed I believe we will solve most of the problems with very little efforts just by splitting these sub templates.

Do you agree this sub-templating will bring value to the maintainability of the css/html?

**Comment by dsantos** [comment]
If you want to keep it simple and not use not use any frontend framework either for styles or rendering, then I agree that templates will be better to what's there now.

**Pull request closed without merging**